Add minting/burning to constructTransaction and decodeTransaction

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

Accommodates https://github.com/cardano-foundation/CIPs/tree/master/CIP-1855
Depends on https://github.com/input-output-hk/cardano-addresses/pull/172
(KeyHash supports CIP 1855 there and handle proper bech32 prefixes)

In this PR we want to have a way to set `txMintValue` of cardano-api's 
```
data TxBodyContent build era =
     TxBodyContent {
       txIns            :: TxIns build era,
      ...
       txMintValue      :: TxMintValue    build era,
       txScriptValidity :: TxScriptValidity era
     }
     deriving (Eq, Show)
```
which inside is
```
data TxMintValue build era where
     TxMintNone  :: TxMintValue build era
     TxMintValue :: MultiAssetSupportedInEra era
                 -> Value
                 -> BuildTxWith build
                      (Map PolicyId (ScriptWitness WitCtxMint era))
                 -> TxMintValue build era
```

This means we need to set both `Value` and `ScriptWitness`. `Value` was already translated to/from `TokenMap`. We need to be careful as cardano-api's `Value` in Integer and in wallet we have two representation for burning/minting, respectively, as they are non-negative. So one of the tasks is to have translation `Script KeyHash` <-> `ScriptWitness`.
Also, we need `(verKey, signKey)` Shelley keys which is regulated in CIP 1855. There is separate derivation path, using purpose=1855 and policy index that is on the same level as account index. Because this is separate key tree `public policy key` was added to seq state (with all consequences to DB, migration). Also separate endpoint for getting the public policy key was exposed (needed in integration testing proving the solution is correct). 

The simplification taken,which does not restrict capabilities, is that we allow policy `ix = 0`. This simplifies storing of policy xpub and discovery, and later delivering signing key. Supporting any policy ix would impose additional complexity, unecessary in my opinion, as we would need to add discovery, pool gaps, extend sequential state, etc. I do think handling `ix=0` only is enough and does not curtail functionality at all. 

Some details about computing `key hash` is that it is `blake2b_224 policyPublicKey`.
Having key hash participating we can determine `policyscript`. It can be any simple script (so nested, with timelocks, multiparty - although not in shelley mode - this will be the case fo multisig). Example od policy scripts.
```
policyKey   -> only signing key is needed
all [policyKey, active_from 120]   -> signing key is needed plus submission must be after slot 120
all [policyKey, active_from 120, active_until 200]  -> signing key is needed plus submission must be after slot 120 but not pass slot 200
```
(now in api we are using cosigner#0 to be compliant with multisig) and it means "us".

We calculate using`policyid = blake2b224 . cbor $ policyscript`

In shelley directory a number of conversions were added that allow translating from api types to node types. Please notice that sneaking `TokenMap` in transaction layer is not enough. `Map TokenPolicyId (Script KeyHash)` is needed. And it was done.

In order to have a prove we have `constructTransaction` working properly we verify tx serialised using `decodeTransaction`. It is reflected in integration testing. 

Please notice we need public policy key to construct any policy script. This has consequences/limitiation of the current impl:
1. We cannot construct this kind of transaction when we are in wallet constructed ONLY from account public key.
2. We can construct this kind of transaction only when in wallet constructed from mnemonic (in that case policy publickey could be derived)
3. Remember that this ticket is part of new tx workflow. so decomposing tx into 3 parts : construct/sign/submit. In that case during construct we are not specifying passwd. 
4. In order to support this kind of transaction in wallet derived from acctXPub we would need to allow optionally to also specify policyXPub side by side acctXPub - to be decided in next PRs
5. There is problem with migrations. I set it to Nothing. And when in mnemonic based wallet with db we cannot enable it via migrations. We would need to add postPolicyXPub  endpoint to set it for wallets that are not from scratch and from mnemonics.  - to be decided in next PRs  


**TO_DO in next PR**: 
1. signing/submitting/showing in integration testing minting/burning works. Here in signing only policyXPub dual, so policyXPrv, should be enough to do the trick (on top of other witnesses for inputs/...)
2. add postPolicyxPub to set policyXPub in Db - also change error msg Piotr got in testnet verification:
```
It seems the wallet lacks policy public key and minting/burningtransaction cannot be realized. Please create wallet from mnemonicsrather than account public key.
```
4. extend `create wallet from acctXPub` endpoint by adding optional policyxPub field. and show this wallet can mint/burn like wallet create from mnemonic 
5. Fail on `construct` while burning assets in case there are no assets of given type to burn.
6. Address boundary values and unpend tests -> https://github.com/input-output-hk/cardano-wallet/pull/3152/commits/a299d67945bd7fca6fc469ba2d863ba08e8ba2e0
7. Address pending integration tests added by Piotr
8. Look if simultaneous minting and burning works ok in Compatibility plus add integration test

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->
ADP-1191

paweljakubas committed:
change data request data for minting/burning asset

paweljakubas committed:
more refactoring and swagger update

paweljakubas committed:
correct core unit tests

paweljakubas committed:
update ApiConstructTransactionData, swagger and core unit tests

paweljakubas committed:
update ApiConstructTransaction, swagger and core unit tests

paweljakubas committed:
add CreatedWrongPolicyScriptTemplate error in constructTransaction

paweljakubas committed:
add integration testing for script template error - part 1

paweljakubas committed:
add integration testing for script template error - part 2

paweljakubas committed:
introduce unimplemented what verkeys are specified

paweljakubas committed:
extend SeqState and add policyXPub

paweljakubas committed:
add migration

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -3323,14 +3345,71 @@ components:
         withdrawal: *transactionWithdrawalRequestSelf
         metadata: *transactionMetadata
-        mint: *transactionMint
+        minted_burned:
```
```suggestion
        mint_burn:
```
I think `mint_burn` would be more consistent with other parameter names in this payload (which don't use past tense), like `payments` or `withdrawal`.

jonathanknowles replied:
I agree with @piotr-iohk here, as `minted`/`burned` seems to imply the past tense, but this looks as though it is for a transaction that has not yet been constructed?

In any case, I think it's clearer to use `mint` and `burn` here.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -2932,6 +2933,106 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+                           [ "cosigner#0",
+                             { "active_from": 120 }
+                           ]
+                        },
```
I wonder if we shouldn't also list the cosigners, like in the shared wallet's `*_script_templates`:
![Screenshot from 2022-02-23 16-56-28](https://user-images.githubusercontent.com/42900201/155356620-e9ce4c79-9d54-4532-91c3-f97d2597e8bf.png)

piotr-iohk replied:
Ok, I guess this comment is not relevant since there can be only one cosigner, but some thoughts on `policy_script_template` in https://github.com/input-output-hk/cardano-wallet/pull/3152#discussion_r824608973.

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -2407,6 +2407,23 @@ components:
+    ApiMintedBurnedInfo: &ApiMintedBurnedInfo
+      type: object
+      required:
+        - verification_key_index
+        - policy_id
+        - asset_name
+        - subject
+        - policy_script
```
I guess that `policy_id`, `subject` and `policy_script` should not be there based on the rendered payload?

![Screenshot from 2022-02-23 16-52-32](https://user-images.githubusercontent.com/42900201/155355987-9374eec6-0aab-4825-b5f7-c8a766eda4f1.png)

paweljakubas committed:
add readPolicyPublicKey

paweljakubas committed:
towards script minting burning - part 1

paweljakubas committed:
towards script minting burning - part 2

paweljakubas committed:
get all info needed from api call needed in shelley

paweljakubas committed:
adapt TransactionCtx to have scripts

paweljakubas committed:
accomodate minting/burning in constructUnsignedTx

paweljakubas committed:
work on Cardano.txMintValue - part 1

paweljakubas committed:
work on Cardano.txMintValue - part 2

paweljakubas committed:
add getPolicyKey endpoint - part 1

paweljakubas committed:
add getPolicyKey endpoint - part 2

paweljakubas committed:
first integration test constructTransaction - decodeTransaction works

paweljakubas committed:
introduce TokenMapWithScripts

paweljakubas committed:
handle ledger script translation for mary

paweljakubas committed:
support timelocks for alonzo and better code reuse

paweljakubas committed:
burning integration testing

paweljakubas committed:
update swagger, api and integration testing

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -2932,6 +2956,251 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+            (Link.decodeTransaction @'Shelley wa) Default decodePayload
+        verify rDecodedTx
+            [ expectResponseCode HTTP.status202
+            , expectField #assetsMinted (`shouldBe` activeAssetsInfo)
```
This decodes to:
```
  "assets_minted": {
    "token_map": [
      {
        "asset_name": "ab12",
        "quantity": 5000,
        "policy_id": "a02c3df561a506bbd265e9dcc98822163fe94b316bccd7f0a22785a9"
      }
    ],
    "policy_scripts": [
      [
        "a02c3df561a506bbd265e9dcc98822163fe94b316bccd7f0a22785a9",
        {
          "all": [
            "addr_shared_vkh1czedspa44x26y07t95uk729cah9rspkta3ry5n7hyry2jefsjw3", <-----------------------------
            {
              "active_from": 120
            }
          ]
        }
      ]
    ],
    "wallet_policy_key_hash": "policy_vkh1czedspa44x26y07t95uk729cah9rspkta3ry5n7hyry2j5mypep"
  }
```
`addr_shared_vkh...` is wrong I believe :thinking:

paweljakubas replied:
indeed, thanks to cardano-adresses 3.9.0 there will be everywhere `policy_vkh` according to standard in CIP 1855

piotr-iohk commented on specifications/api/swagger.yaml:
```diff
@@ -3323,14 +3360,66 @@ components:
+        - operation
+        - policy_script_template
+      properties:
+        policy_script_template: *ScriptTemplateValue
```
Just some thoughts on `policy_script_template`. Currently it looks like this:
![Screenshot from 2022-03-11 11-41-48](https://user-images.githubusercontent.com/42900201/157852252-75b233b4-2552-4451-a0db-3f8d8e0d20fa.png)

So the valid templates could be:
`cosigner#0`
`"all": ["cosigner#0"]`
`"all": ["cosigner#0", { "active_from": 120 }]`
`"all": ["cosigner#0", { "active_from": 120 }, { "active_until": 140 }]`
`"all": ["cosigner#0", any[{ "active_from": 120 }, { "active_until": 140 }]]`

The thing is, you cannot really put anything other than `cosigner#0` and you cannot put more cosigners in the Shelley wallet context. 

The only relevant part is timelocking. Perhaps more user friendly approach would be to leave only timelocking part in the API (which would be non-required) and supply it to some general script internally, e.g.

`"timelocking:" [{ "active_from": 120 }]`
`"timelocking:" [{ "active_from": 120 }, { "active_until": 140 }]`
`"timelocking:" [any[{ "active_from": 120 }, { "active_until": 140 }]]`

If timlocking is empty then it would internally run: `cosigner#0` and if timelocking is added then it would internally run `"all": ["cosigner#0", timelocking_script_part]`

paweljakubas replied:
yes, for shelley mode it could do the trick, but I would not like to be inconsistent with what is for multisig. Both have `Script Cosigner` . Also, it is not small work to have `timelocking` slimmed script handled (slimmed script:= without RequireSignature constructor). There will be another job to merge those, ensure translation between stripts, align with what ApiDecodedTransaction serves, etc. So I am not in favor of introducing timelocking slim script PLUS cosigner#0 instead of normal script used (which is tested and works already). Small user benefit, huge work/maintanace burden

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -2932,6 +2956,251 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+        rTx <- request @(ApiConstructTransaction n) ctx
+            (Link.createUnsignedTransaction @'Shelley wa) Default payload
+        verify rTx
+            [ expectResponseCode HTTP.status202
```
I think burning should fail in this case with the error message saying that there are no tokens to burn. This transaction would fail anyway on sign/submit since there are no tokens with wallet's `policy_id.asset_name` on wallet's balance.

paweljakubas replied:
of course it would fail after submission. this is construction phase. I would not outlaw it in this PR as it is not small. But we can think about this later.

piotr-iohk replied:
Just to note, it is similar case as if you try to construct transaction from an empty wallet... you'll get `not_enough_money` error in that case. The same behavior should be applied here I think. If you try to construct burning tx on the wallet that has nothing to burn you should get proper error on construction phase.

paweljakubas committed:
accommodate cardano-addresses 3.9.0 and adjust code

Commit:
Regenerate nix

paweljakubas committed:
adjust shelley core unit tests

jonathanknowles reviewed (changes requested):
Hi Pawel

Thanks so much for making this PR! Clearly a great deal of work has gone into this.

I'm keen to review it, but before I do that, there is something I'd like to ask:

This PR adds many lines that are far longer than our **80-character line limit**:
https://input-output-hk.github.io/adrestia/code/Coding-Standards#limit-line-length-to-80-characters

Of course, there are some circumstances where it's hard to stay within the limit (for example, in tests for malformed data), but I feel that in the implementation code, we should really try to follow this limit where practical.

Would you be able to address this? There are some strategies listed here:
https://input-output-hk.github.io/adrestia/code/Coding-Standards#strategy-1-wrap-code

Here's a script that can identify long lines:
https://gist.github.com/jonathanknowles/95309c350530c9cebe8eda5d7fd6f3a3

The alternative would be for me to go through this PR and add many, many suggestions, but this would add a lot of noise to the review, which I'm sure you'd agree would distract us from reviewing the actual content!

Many thanks!

Jonathan

paweljakubas committed:
unit test greening and code styling

jonathanknowles committed:
Wrap lines according to team coding standard.

In particular:

    - where practical, restrict lines 80 column lines
    - use 4-space indents everywhere

paweljakubas committed:
fix benchmarks

hlint

jonathanknowles committed:
Fix indentation according to team coding standard.

Fixes:

    - use 4 space indents everywhere.
    - avoid variable-length indentation.

https://input-output-hk.github.io/adrestia/code/Coding-Standards

jonathanknowles committed:
Use consistent terminology for "minted" and "burned".

Use "burned" everywhere, rather than a mixture of "burned" and "burnt".

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2311,39 +2392,67 @@ decodeTransaction
-                map (toApiAnyCert acct acctPath) allCerts
+            map (const (Quantity . fromIntegral . unCoin . W.stakeKeyDeposit $ pp)) $
+            filter ourRewardAccountRegistration $
+            map (toApiAnyCert acct acctPath) allCerts
```
Suggestion: Use the `<$` operator to eliminate `const`:

```suggestion
            (Quantity . fromIntegral . unCoin . W.stakeKeyDeposit $ pp)
                <$ filter ourRewardAccountRegistration
                    (toApiAnyCert acct acctPath <$> allCerts)
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2311,39 +2392,67 @@ decodeTransaction
-                map (toApiAnyCert acct acctPath) allCerts
+            map (const (Quantity . fromIntegral . unCoin . W.stakeKeyDeposit $ pp)) $
+            filter ourRewardAccountDeregistration $
+            map (toApiAnyCert acct acctPath) allCerts
```
Suggestion: Use the `<$` operator to eliminate `const`:
```suggestion
            (Quantity . fromIntegral . unCoin . W.stakeKeyDeposit $ pp)
                <$ filter ourRewardAccountDeregistration
                    (toApiAnyCert acct acctPath <$> allCerts)
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4202,6 +4336,22 @@ instance IsServerError ErrReadRewardAccount where
+            apiError err403 InvalidWalletType $ mconcat
+                [ "It seems the wallet lacks policy public key and minting/burning"
+                , "transaction cannot be realized. Please create wallet from mnemonics"
+                , "rather than account public key."
```
```suggestion
            apiError err403 InvalidWalletType $ unwords
                [ "It seems the wallet lacks a policy public key. It's"
                , "therefore not possible to create a minting or burning"
                , "transaction. Please restore the wallet from a mnemonic"
                , "phrase instead of an account public key."
```

jonathanknowles commented on lib/core-integration/src/Test/Integration/Framework/TestData.hs:
```diff
@@ -669,6 +670,14 @@ errMsg403MultiaccountTransaction = mconcat
+    [ "It looks like I've created a transaction "
+    , "with a minting/burning policy script that either does not "
+    , "pass validation, contains more than one cosigner or "
+    , "cosigner different than cosigner#0."
```
```suggestion
    [ "It looks like I've created a transaction with a minting/burning "
    , "policy script that either does not pass validation, contains more "
    , than one cosigner, or has a cosigner that is different from cosigner#0."
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2235,6 +2265,56 @@ constructTransaction ctx genChange knownPools getPoolStatus (ApiT wid) body = do
+                                (Map.singleton (Cosigner 0) policyXPub)
+                                tName
+                                amt
+                        _ -> error "getMinting should not be used that way"
```
@paweljakubas is this `error` meant to represent a **_programming error_**, or it is possible for the user to trigger it from a faulty API call?

In any case, suggest changing the text to:

```suggestion
                        _ -> error "getMinting should not be used in this way"
```

paweljakubas replied:
it is just impossible  type error in a sense that getMinting is used for values filtered to be minting. Corrected the wording.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2235,6 +2265,56 @@ constructTransaction ctx genChange knownPools getPoolStatus (ApiT wid) body = do
+                                (Map.singleton (Cosigner 0) policyXPub)
+                                tName
+                                amt
+                        _ -> error "getBurning should not be used that way"
```
```suggestion
                        _ -> error "getBurning should not be used in this way"
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2252,12 +2332,13 @@ constructTransaction ctx genChange knownPools getPoolStatus (ApiT wid) body = do
             { transaction = ApiT tx
             , coinSelection = mkApiCoinSelection
                 (maybeToList deposit) (maybeToList refund) Nothing md sel'
+            , mintedBurned = Nothing
```
Should this always be `Nothing`?

paweljakubas replied:
oh, you are right. I forgot to update this. Of course, when minting/burning occurs it should not be Nothing. Added impl and expectations in mint/burn integration testing showing the case. Thanks for this catch.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -3989,6 +4116,13 @@ instance IsServerError ErrConstructTx where
+            [ "It looks like I've created a transaction "
+            , "with a minting/burning policy script that either does not "
+            , "pass validation, contains more than one cosigner or "
+            , "cosigner different than cosigner#0."
```
```suggestion
            apiError err403 CreatedWrongPolicyScriptTemplate $ unwords
            [ "It looks like I've created a transaction with a minting/burning"
            , "policy script that either does not pass validation, contains more"
            , than one cosigner, or has a cosigner that is different from cosigner#0."
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4202,6 +4336,22 @@ instance IsServerError ErrReadRewardAccount where
+            apiError err403 InvalidWalletType $ mconcat
+                [ "It is regrettable but you've just attempted an operation "
+                , "that is invalid for this type of wallet. Only new 'Shelley' "
+                , "wallets can do something with rewards and this one isn't."
```
```suggestion
                [ "You have attempted an operation that is invalid for this "
                , "type of wallet. Only wallets from the Shelley era onwards "
                , "can have rewards, but this wallet is from an era before "
                , "Shelley."
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -2048,6 +2082,42 @@ instance FromJSON ApiVerificationKeyShelley where
+            _ -> fail errRole
+          where
+            errRole =
+                "Unrecognized human-readable part. Expected one of:\
```
```suggestion
                "Unrecognized human-readable part. Expected either\
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -3954,21 +4025,31 @@ instance EncodeAddress n => ToJSON (PostMintBurnAssetData n) where
--- derivation index, which we use to derive the signature to construct the
--- policy with.
+-- circumstances a token may be minted and burned. The policy is the hash of a
+-- serialized script that contain verification keys and timelocks combined in
```
```suggestion
-- serialized script that contains verification keys and timelocks combined in
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -3954,21 +4025,31 @@ instance EncodeAddress n => ToJSON (PostMintBurnAssetData n) where
+-- conditions, possibly nested, to accommodate non-trivial time conditions.
+-- In non-multisig case the script regulating minting/burning will contain
+-- a verification key via cosigner#0 of the wallet with the optional time
+-- predicates.
```
```suggestion
-- In the non-multisig case the script regulating minting/burning will
-- contain a verification key via cosigner#0 of the wallet with optional
-- time predicates.
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -3954,21 +4025,31 @@ instance EncodeAddress n => ToJSON (PostMintBurnAssetData n) where
+-- verification keys of signers (via cosigner#N) with optional time predicates.
+-- The used key derivation index is the same for all engaged derivation keys and
+-- ix=0 is assumed to be used. The verification key derivation is according to
+-- CIP 1855.
```
```suggestion
-- In the multisig case the script regulating minting/burning will contain
-- verification keys of signers (via cosigner#N) with optional time predicates.
-- The used key derivation index is the same for all engaged derivation keys and
-- ix=0 is assumed to be used. The verification key derivation is performed
-- according to CIP 1855.
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -2199,17 +2220,50 @@ instance Arbitrary StakeAddress where
         <$> arbitrary
         <*> arbitrary
         <*> arbitrary
+        <*> arbitrary
```
```suggestion
    arbitrary = applyArbitrary4 ApiConstructTransaction
```

paweljakubas replied:
done already

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1428,26 +1453,42 @@ fromMaryTx tx =
+        = (Map.empty, scriptMap)
+        | (assetsToMint /= TokenMap.empty && assetsToBurn == TokenMap.empty)
+        = (scriptMap, Map.empty)
+        | otherwise = (Map.empty, Map.empty)
```
@paweljakubas What if the `mint` field contains **_both_** assets to mint **_and_** assets to burn?

It seems that in this case, we'd fall through to the `otherwise` case.

Is that expected?

According to my understanding, it's possible to **_both_** mint **_and_** burn assets in a single transaction:
- Assets to **_mint_** are represented by **_positive_** values in the `mint` map.
- Assets to **_burn_** are represented by **_negative_** values in the `mint` map.

The `fromLedgerMintValue` function separates these out into two `TokenMap` values, because `TokenMap` values can only contain non-negative quantities.

paweljakubas replied:
added to investigate that in next PR:
see point 8 in description of this PR : `Look if simultaneous minting and burning works ok in Compatibility plus add integration test`

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:
```diff
@@ -1514,10 +1556,36 @@ fromAlonzoTxBodyAndAux bod mad =
+        = (Map.empty, scriptMap)
+        | (assetsToMint /= TokenMap.empty && assetsToBurn == TokenMap.empty)
+        = (scriptMap, Map.empty)
+        | otherwise = (Map.empty, Map.empty)
```
@paweljakubas What if the `mint` field contains **_both_** assets to mint **_and_** assets to burn?

It seems that in this case, we'd fall through to the `otherwise` case.

Is that expected?

According to my understanding, it's possible to **_both_** mint **_and_** burn assets in a single transaction:
- Assets to **_mint_** are represented by **_positive_** values in the `mint` map.
- Assets to **_burn_** are represented by **_negative_** values in the `mint` map.

The `fromLedgerMintValue` function separates these out into two `TokenMap` values, because `TokenMap` values can only contain non-negative quantities.

paweljakubas replied:
added to investigate that in next PR:
see point 8 in description of this PR : `Look if simultaneous minting and burning works ok in Compatibility plus add integration test`

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -3323,14 +3360,66 @@ components:
+          type: array
+          items: *ApiMintBurnData
+          minItems: 1
+          description: An entry for each unique asset minted and/or burned, containing helpful information.
```
```suggestion
          description: -
              An entry for each unique asset to be minted and/or burned,
              containing helpful information.
```

paweljakubas replied:
done

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -4062,6 +4082,17 @@ x-errCreatedMultiaccountTransaction: &errCreatedMultiaccountTransaction
+  properties:
+    message:
+      type: string
+      description: Occurs when a policy script template is either does not pass validation or has more than one cosigner.
```
```suggestion
      description: -
          Occurs when a policy script template either does not pass validation
          or has more than one cosigner.
```

paweljakubas replied:
done

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -6050,6 +6092,22 @@ paths:
+        <p align="right">status: <strong>stable</strong></p>
+        Return a policy key for a derivation index = 0.
+
+        To get a hash of the policy key, instead of the policy key,
```
```suggestion
        To get a hash of the policy key instead of the policy key,
```

paweljakubas replied:
done

jonathanknowles reviewed (commented):
Hi Pawel

I've finished an initial review of the code. Thanks for making this PR!

I do have a question about what happens when a transaction contains **_both_** assets to mint **_and_** assets to burn. (See comments.).

jonathanknowles committed:
Use `<$` to eliminate `const`.

jonathanknowles committed:
Use `applyArbitrary4` to simplify `Arbitrary` instance.

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -2417,6 +2439,11 @@ components:
         fee: *amount
+        minted_burned:
+          type: array
+          items: *ApiMintedBurnedInfo
```
```suggestion
          items: *ApiMintBurnInfo
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -917,6 +920,7 @@ data ByronWalletPutPassphraseData = ByronWalletPutPassphraseData
 data ApiConstructTransaction (n :: NetworkDiscriminant) = ApiConstructTransaction
     { transaction :: !(ApiT SealedTx)
     , coinSelection :: !(ApiCoinSelection n)
+    , mintedBurned :: !(Maybe (NonEmpty (ApiT ApiMintedBurnedInfo)))
```
```suggestion
    , mintBurn :: !(Maybe (NonEmpty (ApiT ApiMintBurnInfo)))
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1238,7 +1257,8 @@ data ApiDecodedTransaction (n :: NetworkDiscriminant) = ApiDecodedTransaction
 -- | The response cardano-wallet returns upon successful submission of a
 -- mint/burn transaction.
-data ApiMintedBurnedTransaction (n :: NetworkDiscriminant) = ApiMintedBurnedTransaction
+data ApiMintedBurnedTransaction (n :: NetworkDiscriminant) =
```
```suggestion
data ApiMintBurnTransaction (n :: NetworkDiscriminant) =
```

paweljakubas replied:
this one is not used in new tx workflow and should be removed/cleaned in coming PR

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1249,19 +1269,25 @@ data ApiMintedBurnedTransaction (n :: NetworkDiscriminant) = ApiMintedBurnedTran
     deriving anyclass NFData
 
 data ApiMintedBurnedInfo = ApiMintedBurnedInfo
```
```suggestion
data ApiMintBurnInfo = ApiMintBurnInfo
```

paweljakubas replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1219,15 +1223,30 @@ data ApiAnyCertificate n =
+    deriving (Eq, Generic, Show)
+    deriving anyclass NFData
+
+data ApiAssetMintedBurned = ApiAssetMintedBurned
```
```suggestion
data ApiAssetMintBurn = ApiAssetMintBurn
```

paweljakubas replied:
done

Commit:
Add integration tests for boundary values on assetName and minting amount

review response

paweljakubas committed:
some small change to sign properly

Commit:
fixup! Regenerate nix

jonathanknowles committed:
Revise wording of `InvalidWalletType` error.

jonathanknowles committed:
Wrap lines according to team coding standard.

jonathanknowles committed:
Use `MintBurn` terminology consistently.

As discussed in the PR review, this commit adjusts our terminology and
types so that `MintBurn` is used everywhere.

jonathanknowles committed:
Fix `hlint` suggestions to remove redundant `$` operators.

jonathanknowles committed:
Fix `flake.lock` file.

jonathanknowles committed:
Fix CLI specification tests.

sevanspowell reviewed (approved):
This looks sensible to me :+1:

paweljakubas committed:
fix TRANS_NEW_DECODE and more cleaning

jonathanknowles committed:
Wrap long lines according to team coding standard.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles reviewed (approved):
Hi @paweljakubas 

I think this is fine for the moment: given that you plan to add support for transactions that **_both_** mint **_and_** burn in a later PR. 👍🏻 

Many thanks again for making this PR!
